src/mon/MonClient.h

```diff
@@ -209,31 +209,42 @@ class MonClient : public Dispatcher {
 
   // mon subscriptions
 private:
-  map<string,ceph_mon_subscribe_item> sub_have;  // my subs, and current versions
+  map<string,ceph_mon_subscribe_item> sub_sent; // my subs, and current versions
+  map<string,ceph_mon_subscribe_item> sub_new;  // unsent new subs
   utime_t sub_renew_sent, sub_renew_after;
 
   void _renew_subs();
   void handle_subscribe_ack(MMonSubscribeAck* m);
 
   bool _sub_want(string what, version_t start, unsigned flags) {
-    if (sub_have.count(what) &&
-	sub_have[what].start == start &&
-	sub_have[what].flags == flags)
+    if ((sub_new.count(what) == 0 &&
+	 sub_sent.count(what) &&
+	 sub_sent[what].start == start &&
+	 sub_sent[what].flags == flags) ||
+	(sub_new.count(what) &&
+	 sub_new[what].start == start &&
+	 sub_new[what].flags == flags))
       return false;
-    sub_have[what].start = start;
-    sub_have[what].flags = flags;
+    sub_new[what].start = start;
+    sub_new[what].flags = flags;
     return true;
   }
   void _sub_got(string what, version_t got) {
-    if (sub_have.count(what)) {
-      if (sub_have[what].flags & CEPH_SUBSCRIBE_ONETIME)
-	sub_have.erase(what);
+    if (sub_new.count(what)) {
+      if (sub_new[what].flags & CEPH_SUBSCRIBE_ONETIME)
```

Comment:
Is it okay that we're dropping all onetime subscriptions, regardless of the requested epoch values?